035-ca-abstraction.md

```diff
@@ -0,0 +1,396 @@
+In fact Strimzi has two such CAs.
+The "cluster CA" is used for issuing certificates for Strimzi components, such as ZooKeeper and Kafka brokers.
+The "clients CA" is used for issuing certificates for Kafka clients, for example via the User Operator.
+
+The cluster CA root certificate is added to Strimzi component trust stores (so that, for example, ZooKeeper nodes trust each other's and brokers' certificates, and brokers trust certs issued to other brokers, etc). 
+
+
+![Cluster CA relationships](./images/035-cluster-ca.svg)
+
+> In the diagram, the red lines show trust.
+> For example, because the Kafka Broker trusts the Cluster CA certificate, 
+> and the Cluster CA certificate signed the ZK EE certificate
+> the broker will trust the ZK EE certificate presented by the ZK node
+> during TLS handshake.
+
+The cluster CA root certificate also needs to be trusted by Kafka clients connecting to the cluster (so that clients trust the broker's EE certificates). 
+The client CA root certificate is added to the broker trust stores too, so that the brokers will trust certificates issued to Kafka client applications. 
+
+![Client CA relationships](./images/035-clients-ca.svg)
+
+This provides the secure-by-default experience, but in addition to a buggy implementation (**TODO bug ref**) it has very signifiant limitations:
+
+1. It assumes that Strimzi is issuing certificates. 
+2. It doesn't really understand more complex certificate paths, i.e. intermediate certificates.
+3. It doesn't really understand that certificate validity is a property of the whole certificate path.
+4. It doesn't manage the removal of trust in old/unused certificates in a timely manner.
+
+## Motivation
+
+Removing the assumption that Strimzi always issues EE certificates, using a CA certificate it has access to, would enable much more flexibility in CA handling.
+This will be compatible with commonly deployed certificate management solutions such as [Cert Manager][cmio], or [Vault][vault].
+This would be highly valuable for organisations with compliance requirements with regard to certificates.
+
+## Proposal
+
+It is assumed that we want Strimzi to:
+
+* continue to be secure by default. 
+* not have some mandatory dependency on some other project/product for certificate issuance. 
+
+It follows that Strimzi needs to be able to obtain EE certificates, but should not (necessarily) issue them directly itself.
+
+We therefore need an abstraction for how certificates can be issued. 
+In the default (backward compatible) mode this abtraction would continue to use `openssl`, but the new abstraction would also be compatible with issuance by external systems (or even, in theory, a human operator using tooling of their choice and interacting with well defined `Secrets`). 
+
+### Interface for certificate issuance
+
+```java
+public interface CertificateIssuer {
+
+    byte[] requestCertificate(Subject subject, int requestedValidity) throws CertificateIssuanceException;
```

Comment:
Some backend issuers might be paranoid about how and where the key pairs are generated, so there's a good reason to not assume that it's done within the operator.  The implementation of the interface gets to generate a CSR and private key if required by a particular backend issuer. But that doesn't need to be exposed to the operator code which is trying to orchestrate the issuance and deployment of certificates.

Reply:
Ok. I thought about it a bit - I think it might make the the implementation more complicated. But I guess this might be real useful if we one day move beyond Kube Secrets and it will be important that the key never leaves some KSM etc. So I think this is good design. Thanks. Resolving.